File: heads/python/src/main/python/hydra/dsl/meta/lib/eithers.py
"""Phantom-typed term DSL for the hydra.lib.eithers library."""

from hydra.typed import TypedTerm
from hydra.dsl.meta.phantoms import lib_primitive1 as primitive1, lib_primitive2 as primitive2, lib_primitive3 as primitive3


def bimap(f: TypedTerm, g: TypedTerm, e: TypedTerm) -> TypedTerm:
    """Map over both sides of an Either value."""
    return primitive3(f, g, e)


def bind(e: TypedTerm, f: TypedTerm) -> TypedTerm:
    """Bind (flatMap) for Either: if Right, apply the function; if Left, return unchanged."""
    return primitive2(e, f)


def either(f: TypedTerm, g: TypedTerm, e: TypedTerm) -> TypedTerm:
    """Eliminate an Either value by applying one of two functions."""
    return primitive3(f, g, e)


def foldl(f: TypedTerm, init: TypedTerm, xs: TypedTerm) -> TypedTerm:
    """Left-fold over a list with an Either-returning function, short-circuiting on Left."""
    return primitive3(f, init, xs)


def from_left(default: TypedTerm, e: TypedTerm) -> TypedTerm:
    """Extract the Left value, or return a default."""
    return primitive2(default, e)


def from_right(default: TypedTerm, e: TypedTerm) -> TypedTerm:
    """Extract the Right value, or return a default."""
    return primitive2(default, e)


def is_left(e: TypedTerm) -> TypedTerm:
    """Check if an Either is a Left value."""
    return primitive1(e)


def is_right(e: TypedTerm) -> TypedTerm:
    """Check if an Either is a Right value."""
    return primitive1(e)


def lefts(eithers: TypedTerm) -> TypedTerm:
    """Extract all Left values from a list of Eithers."""
    return primitive1(eithers)


def map_(f: TypedTerm, e: TypedTerm) -> TypedTerm:
    """Map a function over the Right side of an Either (standard functor map)."""
    return primitive2(f, e)


def map_list(f: TypedTerm, xs: TypedTerm) -> TypedTerm:
    """Map a function returning Either over a list, collecting results or short-circuiting on Left."""
    return primitive2(f, xs)


def map_maybe(f: TypedTerm, x: TypedTerm) -> TypedTerm:
    """Map a function returning Either over a Maybe, or return Right Nothing if Nothing."""
    return primitive2(f, x)


def map_set(f: TypedTerm, s: TypedTerm) -> TypedTerm:
    """Map a function returning Either over a Set, collecting results or short-circuiting on Left."""
    return primitive2(f, s)


def partition_eithers(eithers: TypedTerm) -> TypedTerm:
    """Partition a list of Eithers into lefts and rights."""
    return primitive1(eithers)


def rights(eithers: TypedTerm) -> TypedTerm:
    """Extract all Right values from a list of Eithers."""
    return primitive1(eithers)

File: heads/haskell/src/main/haskell/Hydra/Dsl/Meta/Lib/Eithers.hs
-- | Phantom-typed term DSL for the hydra.lib.eithers library

module Hydra.Dsl.Meta.Lib.Eithers where

import Hydra.Typed
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Sources.Libraries
import qualified Data.Set as S


-- | Map over both sides of an Either value.
bimap :: TypedTerm (a -> c) -> TypedTerm (b -> d) -> TypedTerm (Either a b) -> TypedTerm (Either c d)
bimap = primitive3 _eithers_bimap

-- | Bind (flatMap) for Either: if Right, apply the function; if Left, return unchanged.
bind :: TypedTerm (Either a b) -> TypedTerm (b -> Either a c) -> TypedTerm (Either a c)
bind = primitive2 _eithers_bind

-- | Eliminate an Either value by applying one of two functions.
either_ :: TypedTerm (a -> c) -> TypedTerm (b -> c) -> TypedTerm (Either a b) -> TypedTerm c
either_ = primitive3 _eithers_either

-- | Left-fold over a list with an Either-returning function, short-circuiting on Left.
foldl :: TypedTerm (a -> b -> Either c a) -> TypedTerm a -> TypedTerm [b] -> TypedTerm (Either c a)
foldl = primitive3 _eithers_foldl

-- | Extract the Left value, or return a default.
fromLeft :: TypedTerm a -> TypedTerm (Either a b) -> TypedTerm a
fromLeft = primitive2 _eithers_fromLeft

-- | Extract the Right value, or return a default.
fromRight :: TypedTerm b -> TypedTerm (Either a b) -> TypedTerm b
fromRight = primitive2 _eithers_fromRight

-- | Check if an Either is a Left value.
isLeft :: TypedTerm (Either a b) -> TypedTerm Bool
isLeft = primitive1 _eithers_isLeft

-- | Check if an Either is a Right value.
isRight :: TypedTerm (Either a b) -> TypedTerm Bool
isRight = primitive1 _eithers_isRight

-- | Extract all Left values from a list of Eithers.
lefts :: TypedTerm [Either a b] -> TypedTerm [a]
lefts = primitive1 _eithers_lefts

-- | Map a function over the Right side of an Either (standard functor map).
map :: TypedTerm (a -> b) -> TypedTerm (Either c a) -> TypedTerm (Either c b)
map = primitive2 _eithers_map

-- | Map a function returning Either over a list, collecting results or short-circuiting on Left.
mapList :: TypedTerm (a -> Either e b) -> TypedTerm [a] -> TypedTerm (Either e [b])
mapList = primitive2 _eithers_mapList

-- | Map a function returning Either over an optional, or return Right Nothing if Nothing.
mapOptional :: TypedTerm (a -> Either c b) -> TypedTerm (Maybe a) -> TypedTerm (Either c (Maybe b))
mapOptional = primitive2 _eithers_mapOptional

-- | Map a function returning Either over a Set, collecting results or short-circuiting on Left.
mapSet :: TypedTerm (a -> Either c b) -> TypedTerm (S.Set a) -> TypedTerm (Either c (S.Set b))
mapSet = primitive2 _eithers_mapSet

-- | Partition a list of Eithers into lefts and rights.
partitionEithers :: TypedTerm [Either a b] -> TypedTerm ([a], [b])
partitionEithers = primitive1 _eithers_partitionEithers

-- | Extract all Right values from a list of Eithers.
rights :: TypedTerm [Either a b] -> TypedTerm [b]
rights = primitive1 _eithers_rights
